starlette/testclient.py

```diff
@@ -452,42 +446,56 @@ def websocket_connect(
         return session
 
     def __enter__(self) -> "TestClient":
-        loop = asyncio.get_event_loop()
-        self.send_queue = asyncio.Queue()  # type: asyncio.Queue
-        self.receive_queue = asyncio.Queue()  # type: asyncio.Queue
-        self.task = loop.create_task(self.lifespan())
-        loop.run_until_complete(self.wait_startup())
+        self.stream_send, self.stream_receive = anyio.create_memory_object_stream()
+        self.portal = anyio.start_blocking_portal(
+            backend_options={"debug": True}
+        )  # XXX backend
+        self.task = self.portal.spawn_task(self.lifespan)
+        self.portal.call(self.wait_startup)
         return self
 
     def __exit__(self, *args: typing.Any) -> None:
-        loop = asyncio.get_event_loop()
-        loop.run_until_complete(self.wait_shutdown())
+        try:
+            self.portal.call(self.wait_shutdown)
+        finally:
+            self.portal.stop_from_external_thread()
 
     async def lifespan(self) -> None:
         scope = {"type": "lifespan"}
-        try:
-            await self.app(scope, self.receive_queue.get, self.send_queue.put)
-        finally:
-            await self.send_queue.put(None)
+        async with self.stream_send:
+            await self.app(scope, self.stream_receive.receive, self.stream_send.send)
 
     async def wait_startup(self) -> None:
-        await self.receive_queue.put({"type": "lifespan.startup"})
-        message = await self.send_queue.get()
-        if message is None:
+        try:
+            await self.stream_send.send({"type": "lifespan.startup"})
+        except anyio.ClosedResourceError:
+            self.task.result()
+            return
+        try:
+            message = await self.stream_receive.receive()
+        except anyio.EndOfStream:
             self.task.result()
+            return
         assert message["type"] in (
             "lifespan.startup.complete",
             "lifespan.startup.failed",
         )
         if message["type"] == "lifespan.startup.failed":
-            message = await self.send_queue.get()
-            if message is None:
+            try:
+                message = await self.stream_receive.receive()
+            except anyio.EndOfStream:
                 self.task.result()
 
     async def wait_shutdown(self) -> None:
-        await self.receive_queue.put({"type": "lifespan.shutdown"})
-        message = await self.send_queue.get()
-        if message is None:
+        try:
+            await self.stream_send.send({"type": "lifespan.shutdown"})
+        except anyio.ClosedResourceError:
+            self.task.result()
+            return
+        try:
+            message = await self.stream_receive.receive()
+        except anyio.EndOfStream:
             self.task.result()
+            return
         assert message["type"] == "lifespan.shutdown.complete"
-        await self.task
+        self.task.result()
```

Comment:
What is `self.task` here?

Reply:
I've changed a bit more, but `self.task` should be `self.portal.spawn_task(self.lifespan)`